File: DataStructures/StacksQueues/implementing_queue.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class Queue {
    constructor() {
        this.first = null;
        this.last = null;
        this.length = 0;
    }
    peek() {
        return this.first;
    }
    enqueue(value) {
        let newNode = new Node(value);
        if (!this.last) {
            this.first = this.last = newNode;
            this.length++;
            return this;
        }
        let lastNode = this.last;
        lastNode.next = newNode;
        this.last = newNode;
        this.length++;
        return this;
    }
    dequeue() {
        if (!this.last) {
            return null;
        }
        if (this.length === 1) {
            this.first = this.last = null;
            this.length--;
            return null;
        }
        let firstNode = this.first;
        this.first = firstNode.next;
        this.length--;
        return firstNode;
    }
    isEmpty() {
        return !Boolean(this.length);
    }
}

if (require.main === module) {
    const newQueue = new Queue();
    console.log(newQueue.isEmpty());
    console.log(newQueue.enqueue("abc"));
    console.log(newQueue.isEmpty());
    console.log(newQueue.enqueue(245));
    console.log(newQueue.peek());
    console.log(newQueue.dequeue());
    console.log(newQueue.peek());
}

File: DataStructures/StacksQueues/implementing_queue.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class Queue:
    def __init__(self):
        self.first = None
        self.last = None
        self.length = 0
    
    def peek(self):
        return self.first.value
    
    def enqueue(self, value):
        new_node = Node(value)
        if self.last is None:
            self.first = self.last = new_node
            self.length += 1
            return self
        last_node = self.last
        last_node.next = new_node
        self.last = new_node
        self.length += 1
        return self
    
    def dequeue(self):
        if self.last is None:
            return None
        if self.length == 1:
            self.first = self.last = None
            self.length -= 1
            return None
        first_node = self.first
        self.first = first_node.next
        self.length -= 1
        return first_node
    
    def is_empty(self):
        return not bool(self.length)


if __name__ == "__main__":
    new_queue = Queue()
    print(new_queue.is_empty())
    print(new_queue.enqueue("abc"))
    print(new_queue.is_empty())
    print(new_queue.enqueue(245))
    print(new_queue.peek())
    print(new_queue.dequeue())
    print(new_queue.peek())
